// build-html/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _regenerator = require('babel-runtime/regenerator');

var _regenerator2 = _interopRequireDefault(_regenerator);

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _htmlMinifier = require('html-minifier');

var _directoryHelpers = require('directory-helpers');

var _directoryHelpers2 = _interopRequireDefault(_directoryHelpers);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var project = new _directoryHelpers2.default('.');

exports.default = (0, _asyncToGenerator3.default)(_regenerator2.default.mark(function _callee() {
  var html, minifiedHtml;
  return _regenerator2.default.wrap(function _callee$(_context) {
    while (1) {
      switch (_context.prev = _context.next) {
        case 0:
          _context.next = 2;
          return project.read('src/index.html');

        case 2:
          html = _context.sent;
          minifiedHtml = (0, _htmlMinifier.minify)(html, {
            collapseBooleanAttributes: true,
            collapseWhitespace: true,
            decodeEntities: true,
            includeAutoGeneratedTags: false,
            removeAttributeQuotes: true,
            removeComments: true,
            removeEmptyAttributes: true,
            removeOptionalTags: true,
            removeRedundantAttributes: true,
            removeScriptTypeAttributes: true,
            removeStyleLinkTypeAttributes: true,
            sortAttributes: true,
            sortClassName: true,
            useShortDoctype: true
          });
          _context.next = 6;
          return project.write({ 'dist/index.html': minifiedHtml });

        case 6:
        case 'end':
          return _context.stop();
      }
    }
  }, _callee, this);
}));
